### PR TITLE
UART for the nRF

### PR DESCRIPTION
**Pull request by utkarshapets**
Setting up UART support for the nRF51822. Involved:
Creating src/chips/nrf51822/uart.rs file

### CODE REVIEW

File: src/chips/nrf51822/gpio.rs [hunk 1]
```diff
@@ -1,103 +0,0 @@
-use core::mem;
```
Comment (alevy) [inline src/chips/nrf51822/gpio.rs]: don't remove this file, just leave it as is in master:

```bash
$ git checkout master -- src/chips/nrf51822/gpio.rs
```

File: src/platform/nrf_pca10001/lib.rs [hunk 2]
```diff
@@ -2,16 +2,35 @@
 #![crate_type = "rlib"]
```
Comment (alevy) [inline src/platform/nrf_pca10001/lib.rs]: There are no changes here that are related to the UART, and many of the changes are incorrectly reverting previous changes. Just leave this file as is.

File: src/platform/nrf_pca10001/lib.rs [hunk 3]
```diff
@@ -40,86 +58,104 @@ impl Platform {
...
-    static_init!(gpio: drivers::gpio::GPIO<'static, nrf51822::gpio::GPIOPin> =
-                     drivers::gpio::GPIO::new(gpio_pins), 20);
+
+#[inline(never)]
+	
+pub unsafe fn init<'a>() -> &'a mut Firestorm {
+
+    use core::mem;
+
+	let mut nrf_uart = uart::UART::new();
```
Comment (alevy) [inline src/platform/nrf_pca10001/lib.rs]: maybe with the exception of these particular changes. You can just add them back in by hand

File: src/platform/nrf_pca10001/systick.rs [hunk 4]
```diff
@@ -1,26 +0,0 @@
-/// Stub implementation of a systick timer since the NRF doesn't have the
```
Comment (alevy) [inline src/platform/nrf_pca10001/systick.rs]: don't remove this file, just leave it as is in master:

```bash
$ git checkout master -- src/platform/nrf_pca10001/systick.rs
```

File: src/chips/nrf51822/peripheral_interrupts.rs [hunk 5]
```diff
@@ -0,0 +1,31 @@
+// Automatically generated by tools/nRF51_codegen.py
```
Comment (alevy) [inline src/chips/nrf51822/peripheral_interrupts.rs]: Why do you need this?

File: apps/Makefile.nrf_pca10001.mk [hunk 6]
```diff
@@ -9,19 +9,24 @@ JLINK_EXE ?= JLinkExe
 
 # Apps to link may grow over time so defer expanding that
 .SECONDEXPANSION:
-$(TOCK_APP_BUILD_DIR)/kernel_and_app.elf: $(TOCK_BUILD_DIR)/ctx_switch.o $(TOCK_BUILD_DIR)/kernel.o $$(APPS_TO_LINK_TO_KERNEL) | $(TOCK_BUILD_DIR)
+$(TOCK_APP_BUILD_DIR)/kernel_and_app.elf: $(TOCK_BUILD_DIR)/ctx_switch.o $(TOCK_BUILD_DIR)/kernel.o $$(TOCK_BUILD_DIR)/kernel.o $$(APPS_TO_LINK_TO_KERNEL) | $(TOCK_BUILD_DIR)
```
Comment (alevy) [inline apps/Makefile.nrf_pca10001.mk]: This seems like it's just adding a duplicate dependency, is this actually necessary?

File: apps/Makefile.nrf_pca10001.mk [hunk 7]
```diff
@@ -56,3 +61,8 @@ erase-all:
 	r\\n\
 	exit | $(JLINK) $(JLINK_OPTIONS)
 
+MOUNT_DIR=/var/run/usbmount/MBED_microcontroller
```
Comment (alevy) [inline apps/Makefile.nrf_pca10001.mk]: How about `MBED_DIR` instead?

File: src/chips/nrf51822/gpio.rs [hunk 8]
```diff
@@ -101,3 +101,4 @@ pub static mut PA : Port = Port {
         GPIOPin::new(28), GPIOPin::new(29), GPIOPin::new(30), GPIOPin::new(31),
     ],
 };
+
```
Comment (alevy) [inline src/chips/nrf51822/gpio.rs]: uh... if you get rid of this extra newline, it's one more file with no changes

File: src/chips/nrf51822/lib.rs [hunk 9]
```diff
@@ -1,10 +1,16 @@
...
 #![crate_type = "rlib"]
 #![feature(asm,concat_idents,const_fn)]
+#![feature(core_intrinsics)]
 #![no_std]
 
 extern crate common;
 extern crate hil;
 
 mod peripheral_registers;
+mod peripheral_interrupts;
```
Comment (alevy) [inline src/chips/nrf51822/lib.rs]: Yeah, just remove this. You don't need `peripheral_interrupts` for your changes.

File: apps/Makefile.nrf_pca10001.mk (hunk 6 above)
Comment (utkarshapets) [inline apps/Makefile.nrf_pca10001.mk]: I don't recall changing this line, so I will have to read the file thoroughly to figure out whether it is needed- I'll do that now

File: apps/Makefile.nrf_pca10001.mk (hunk 7 above)
Comment (utkarshapets) [inline apps/Makefile.nrf_pca10001.mk]: I will have to create something like MBED_DIR, right?

File: src/chips/nrf51822/peripheral_interrupts.rs (hunk 5 above)
Comment (utkarshapets) [inline src/chips/nrf51822/peripheral_interrupts.rs]: Well, this file is created by the nrf51_codegen script. It basically goes over a XML file which describes the platform and creates a listing of the interrupts and registers. It doesn't do so for UART though, so I switched to using the nvic file and writing in the register names myself.
However I am not really using interrupts anyway

File: apps/Makefile.nrf_pca10001.mk (hunk 7 above)
Comment (alevy) [inline apps/Makefile.nrf_pca10001.mk]: I'm just saying rename the variable `MOUNT_DIR` to `MBED_DIR`

**Comment by alevy** [comment]
At this point, worth doing a `git rebase -i master` to clean up the commits so far into one or two meaningful commits.

File: apps/Makefile.nrf_pca10001.mk [hunk 10]
```diff
@@ -17,11 +17,16 @@ $(TOCK_APP_BUILD_DIR)/kernel_and_app.elf: $(TOCK_BUILD_DIR)/ctx_switch.o $(TOCK_
 
 # XXX Temporary until new kernel build system in place
 $(TOCK_BUILD_DIR)/ctx_switch.o: kernel
+$(TOCK_BUILD_DIR)/crt1.o: kernel
```
Comment (alevy) [inline apps/Makefile.nrf_pca10001.mk]: You actually need to depend on `crt1.o` in `kernel_and_app.elf`

File: src/chips/nrf51822/uart.rs [hunk 11]
```diff
@@ -0,0 +1,145 @@
...
+        unsafe{intrinsics::volatile_load(&regs.rxdrdy) & 0b1 != 0 }
+    }
+
+    fn tx_ready(&self) -> bool {
+        let regs : &Registers = unsafe { mem::transmute(self.regs) };
+        unsafe{intrinsics::volatile_load(&regs.txdrdy) & 0b1 != 0}
+    }
+
+    pub fn send_byte(&self, byte: u8) {
+        while !self.tx_ready() {}
```
Comment (alevy) [inline src/chips/nrf51822/uart.rs]: `tx_read` is only meaningful after sending the first byte in a transaction. So the steps are:

  1. Start TX
  2. Send first byte
  3. Wait for TX_Ready
  4. If more byte go back to 2
  5. Stop TX

File: src/chips/nrf51822/uart.rs [hunk 12]
```diff
@@ -0,0 +1,145 @@
...
+
+use core::mem;
+
+//use hil::uart;
+//use nvic;
+//use chip;
+use core::intrinsics;
+
+struct Registers {
+    startrx: u32,
```
Comment (alevy) [inline src/chips/nrf51822/uart.rs]: Replace these field types with `VolatileCell<u32>` (not necessary for the reserved fields) and replace instances of `volatile_load(field)` and `volatile_store(field, val)` with `field.get()` and `field.set(val)`, respectively

**Comment by alevy** [comment]
Superseded by #93

**Pull request closed without merging**